### PR TITLE
usb: set SN string descriptor at runtime

### PR DESCRIPTION
**Pull request by qianfan-Zhao**
implement a method that can set SN string descriptor at runtime.
but runtime SN and default SN configured in Kconfig must has the
same length.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

### CODE REVIEW

**Comment by zhaozhongchen** [comment]
there has a simple example that change SN from "0.01" (configured in Kconfig) to "0.0X" at runtime:

```
diff --git a/samples/subsys/usb/hid-mouse/src/main.c b/samples/subsys/usb/hid-mouse/src/main.c
index ff2a10b..d865d4d 100644
--- a/samples/subsys/usb/hid-mouse/src/main.c
+++ b/samples/subsys/usb/hid-mouse/src/main.c
@@ -124,6 +124,14 @@ void button_pressed(struct device *gpio, struct gpio_callback *cb,
        }
 }
 
+u8_t * usb_set_runtime_ascii_sn_string_desc(void)
+{
+       static u8_t ascii_sn[sizeof(CONFIG_USB_DEVICE_SN)] = {0};
+
+       memcpy(ascii_sn, "0.0X", 4);
+       return ascii_sn;
+}
+
 void main(void)
 {
        struct device *gpio;
```

File: subsys/usb/usb_descriptor.c [hunk 1]
```diff
@@ -818,8 +819,48 @@ void ascii7_to_utf16le(int idx_max, int asci_idx_max, u8_t *buf)
...
+
+	/* both runtime and default SN's length doesn't include \0 */
+	runtime_sn_len = strlen(runtime_sn);
+	default_sn_len = sizeof(CONFIG_USB_DEVICE_SN) - 1;
+
+	if (runtime_sn_len != default_sn_len) {
+		SYS_LOG_ERR("the new SN descriptor doesn't has the same "
+			    "length as CONFIG_USB_DEVICE_SN");
+		for (; ;) {
+			/* Spin */
```
Comment (jfischer-phytec-iot) [inline subsys/usb/usb_descriptor.c]: Why? It should return errors or output warning, it is also not critical to the function of the stack.

File: subsys/usb/usb_descriptor.c [hunk 2]
```diff
@@ -818,8 +819,48 @@ void ascii7_to_utf16le(int idx_max, int asci_idx_max, u8_t *buf)
...
 
+/*
+ * default USB SN string descriptor is CONFIG_USB_DEVICE_SN, but sometimes
+ * we want use another string as SN descriptor such as the chip's unique
+ * ID. So user can implement this function return a string thats will be
+ * replaced the default SN.
+ * please note that the new SN descriptor you changed must has same length
+ * as CONFIG_USB_DEVICE_SN.
+ */
+__weak u8_t *usb_set_runtime_ascii_sn_string_desc(void)
```
Comment (jfischer-phytec-iot) [inline subsys/usb/usb_descriptor.c]: You can do it that way, but name of the function can be more generic, not limited to usb, something like SoC/board serial number API (cc @carlescufi)

**Review by jfischer-no — CHANGES_REQUESTED** [review]
see also #4661

File: subsys/usb/usb_descriptor.c (hunk 1 above)
Comment (zhaozhongchen) [reply subsys/usb/usb_descriptor.c]: yeah it's better leaving a warring message and return, not enter a dead while.

File: subsys/usb/usb_descriptor.c (hunk 2 above)
Comment (finikorg) [reply subsys/usb/usb_descriptor.c]: BTW what is the use case for this? I understand the same could be done with configuration option.

Comment (zhaozhongchen) [reply subsys/usb/usb_descriptor.c]: @finikorg if two same product has same VID and PID, we can't distinguish them. So in that case we need give them a different SN. Many MCU has unique ID, and we can generate USB SN from it's UID at runtime.

There should has a way to rewrite SN descriptor, get SN from EEPROM or generated from UID. Regardless of the method, it needs to be implemented this function.

File: subsys/usb/usb_descriptor.c [hunk 3]
```diff
@@ -818,8 +819,46 @@ void ascii7_to_utf16le(int idx_max, int asci_idx_max, u8_t *buf)
...
+	runtime_sn_len = strlen(runtime_sn);
+	default_sn_len = sizeof(CONFIG_USB_DEVICE_SN) - 1;
+
+	if (runtime_sn_len != default_sn_len) {
+		SYS_LOG_ERR("the new SN descriptor doesn't has the same "
+			    "length as CONFIG_USB_DEVICE_SN");
+		return ;
+	}
+
+	memcpy((u8_t *)common_desc.string_descr.utf16le_sn.bString,
```
Comment (finikorg) [inline subsys/usb/usb_descriptor.c]: Isn't is already u8_t?

File: subsys/usb/usb_descriptor.c [hunk 4]
```diff
@@ -818,8 +819,46 @@ void ascii7_to_utf16le(int idx_max, int asci_idx_max, u8_t *buf)
 	}
 }
 
+/*
+ * default USB SN string descriptor is CONFIG_USB_DEVICE_SN, but sometimes
+ * we want use another string as SN descriptor such as the chip's unique
+ * ID. So user can implement this function return a string thats will be
+ * replaced the default SN.
+ * please note that the new SN descriptor you changed must has same length
```
Comment (finikorg) [inline subsys/usb/usb_descriptor.c]: Please start sentence with capital letter.

**Review by finikorg — COMMENTED** [review]
Generally OK

File: subsys/usb/usb_descriptor.c (hunk 2 above)
Comment (zhaozhongchen) [reply subsys/usb/usb_descriptor.c]: @jfischer-phytec-iot This function just change usb's SN string descriptor, not do other work. what's you idea about the function's name?

File: subsys/usb/usb_descriptor.c (hunk 3 above)
Comment (zhaozhongchen) [reply subsys/usb/usb_descriptor.c]: Yes it is and I had fixed it

File: subsys/usb/usb_descriptor.c (hunk 4 above)
Comment (zhaozhongchen) [reply subsys/usb/usb_descriptor.c]: Done

File: subsys/usb/usb_descriptor.c (hunk 2 above)
Comment (jfischer-phytec-iot) [reply subsys/usb/usb_descriptor.c]: I think about something generic, get_sn_board or get_sn_soc, is not necessarily limited to usb. Fixes: #6593

File: subsys/usb/usb_descriptor.c [hunk 5]
```diff
@@ -818,8 +819,45 @@ void ascii7_to_utf16le(int idx_max, int asci_idx_max, u8_t *buf)
...
+	default_sn_len = strlen(CONFIG_USB_DEVICE_SN);
+
+	if (runtime_sn_len != default_sn_len) {
+		SYS_LOG_ERR("the new SN descriptor doesn't has the same "
+			    "length as CONFIG_USB_DEVICE_SN");
+		return ;
+	}
+
+	memcpy(common_desc.string_descr.utf16le_sn.bString,
+		runtime_sn, runtime_sn_len); /* exclude last '\0' */
```
[the remark this answers is not in the document]
Comment (zhaozhongchen) [reply subsys/usb/usb_descriptor.c]: done

**Comment by zhaozhongchen** [comment]
@jfischer-phytec-iot I had changed that function's name to usb_ascii_sn_descriptor_board_fixup.

**Review by finikorg — APPROVED** [review]

**Comment by carlescufi** [comment]
@jfischer-phytec-iot can you revisit this?

**Comment by jfischer-phytec-iot** [comment]
I would like to keep it open until #4661 is merged.

**Comment by zhaozhongchen** [comment]
@jfischer-phytec-iot @finikorg I had rebased, please recheck

File: subsys/usb/usb_descriptor.c [hunk 6]
```diff
@@ -44,6 +45,10 @@ struct common_descriptor {
 	struct usb_cfg_descriptor cfg_descr;
 } __packed;
 
+#define USB_DESC_iManufacturer				1
```
Comment (jfischer-phytec-iot) [inline subsys/usb/usb_descriptor.c]: change to capital letters and add _IDX

File: subsys/usb/usb_descriptor.c [hunk 7]
```diff
@@ -281,6 +286,41 @@ static struct usb_cfg_data *usb_get_cfg_data(struct usb_if_descriptor *iface)
...
+ * Please note that the new SN descriptor you changed must has same length
+ * as CONFIG_USB_DEVICE_SN.
+ */
+__weak u8_t *usb_ascii_sn_descriptor_board_fixup(void)
+{
+	return NULL;
+}
+
+static void
+usb_fix_ascii_sn_string_descriptor(struct usb_sn_descriptor *sn)
```
Comment (jfischer-phytec-iot) [inline subsys/usb/usb_descriptor.c]: It does not fix anything, it only updates the string,  usb_update_sn_string_descriptor would be more accurate.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
@qianfan-Zhao mind taking a look at the latest review comments?

**Comment by qianfan-Zhao** [comment]
@carlescufi @jfischer-phytec-iot Sorry I had noticed that comments and I haven't has time to update my code. I have a job and I will do my job during the day and only pay attention to Zephyr on my spare time.

I will update this pr tonight.
Thanks yours remind!

**Comment by qianfan-Zhao** [comment]
@jfischer-phytec-iot please recheck

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**